tests/delete/models.py

```diff
@@ -147,3 +158,45 @@ class SecondReferrer(models.Model):
     other_referrer = models.ForeignKey(
         Referrer, models.CASCADE, to_field='unique_field', related_name='+'
     )
+
+
+class DeleteTop(models.Model):
```

Comment:
Do we the second structure with `GenericRelation`? it seems to me that adding ``ParentA`` should be enough to test path without fast-delete:
```diff
class ParentA(models.Model):
    cascade = models.ForeignKey(A, models.CASCADE)
```
also some tests fail when I add this model :thinking: 
```
django.db.utils.IntegrityError: (1451, 'Cannot delete or update a parent row: a foreign key constraint fails (`test_django`.`delete_a`, CONSTRAINT `delete_a_restrict_id_0b88afcb_fk_delete_r_id` FOREIGN KEY (`restrict_id`) REFERENCES `delete_r` (`id`))')
```

Reply:
Fast deletes are being tested with `GenericDeleteBottomRelated` here, which is similar to the `ParentA` you're proposing.

The second structure with `GenericRelation` is not for fast deletes but to specifically test generic relations. Without this test, the entire `delete` test suite passes even if this line is deleted: https://github.com/django/django/blob/4527d5db0f0a939439b9373f7a9a4eee3386afff/django/db/models/deletion.py#L258